### PR TITLE
Fix gMSA locks

### PR DESCRIPTION
**Pull request by vsiddharth**
Signed-off-by: Vinothkumar Siddharth <sidvin@amazon.com>

### Summary

This patch decouples the credentialspec status and data locks.

### Implementation details

The datalock would protect access to the credential spec resources.
The other lock would be used for all other transitions.

### Testing

New tests cover the changes: 
No 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/taskresource/credentialspec/credentialspec_windows.go [hunk 1]
```diff
@@ -98,6 +101,8 @@ func NewCredentialSpecResource(taskARN, region string,
 		ssmClientCreator:        ssmClientCreator,
 		s3ClientCreator:         s3ClientCreator,
 		CredSpecMap:             make(map[string]string),
+		os:                      oswrapper.NewOS(),
```
Comment (yumex93) [inline agent/taskresource/credentialspec/credentialspec_windows.go]: Where are these used?

File: agent/taskresource/credentialspec/credentialspec_windows.go [hunk 2]
```diff
@@ -77,6 +77,9 @@ type CredentialSpecResource struct {
 	// * key := credentialspec:s3ARN, value := credentialspec=file://CredentialSpecResourceLocation/s3_taskARN_fileName.json
 	// * key := credentialspec:ssmARN, value := credentialspec=file://CredentialSpecResourceLocation/ssm_taskARN_param.json
 	CredSpecMap map[string]string
+	// datalock should protect the CredSpecMap
+	datalock sync.RWMutex
```
Comment (sharanyad) [inline agent/taskresource/credentialspec/credentialspec_windows.go]: could you also add comments on why this separate lock is needed?

Comment (vsiddharth) [reply agent/taskresource/credentialspec/credentialspec_windows.go]: I added information to the commit message. 
Do you want it replicated here?

File: agent/taskresource/credentialspec/credentialspec_windows.go (hunk 1 above)
Comment (vsiddharth) [reply agent/taskresource/credentialspec/credentialspec_windows.go]: There are used in the handlers within the package.

Comment (vsiddharth) [reply agent/taskresource/credentialspec/credentialspec_windows.go]: * https://github.com/aws/amazon-ecs-agent/blob/feature/gMSA/agent/taskresource/credentialspec/credentialspec_windows.go#L525
* https://github.com/aws/amazon-ecs-agent/blob/feature/gMSA/agent/taskresource/credentialspec/credentialspec_windows.go#L468

File: agent/taskresource/credentialspec/credentialspec_windows.go (hunk 2 above)
Comment (sharanyad) [reply agent/taskresource/credentialspec/credentialspec_windows.go]: that explains 'what' the locks are for. I'm looking for the answer for 'why'.

Comment (vsiddharth) [reply agent/taskresource/credentialspec/credentialspec_windows.go]: Added more details.

**Comment by fenxiong** [comment]
I think the deadlock is because you get the lock in `ApplyTransition` https://github.com/aws/amazon-ecs-agent/blob/dd6a824a05f9e9fbfd980b32a06f3ae59065b619/agent/taskresource/credentialspec/credentialspec_windows.go#L195-L196 which will then call `Create`, and `Create` will stuck at getting the same lock (might be because Firelens resource does that but Firelens doesn't get lock during `Create`...). I think if you remove the lock in `ApplyTransition` you shouldn't have this issue, and wouldn't need another lock.

**Comment by vsiddharth** [comment]
Superseded by https://github.com/aws/amazon-ecs-agent/pull/2279

**Pull request closed without merging**